app/models/reports/schema_v2.rb

```diff
@@ -0,0 +1,230 @@
+module Reports
+  class SchemaV2
+    include BustCache
+    include Memery
+
+    attr_reader :control_rate_query_v2
+    attr_reader :periods
+    attr_reader :period_hash
+    attr_reader :period_type
+    attr_reader :regions
+
+    delegate :cache, :logger, to: Rails
+
+    def initialize(regions, periods:)
+      @regions = regions
+      @periods = periods
+      @period_type = @periods.first.type
+      @control_rate_query_v2 = ControlRateQueryV2.new
+      @period_hash = lambda { |month_date, count| [Period.month(month_date), count] }
+    end
+
+    # Returns the earliest patient record for a Region from either assigned or registered patients. Note that this *ignores*
+    # the periods that are passed in for the Repository - this is the true 'earliest report date' for a Region.
+    memoize def earliest_patient_recorded_at
+      region_entries = regions.map { |region| RegionEntry.new(region, __method__) }
+      cached_results = cache.fetch_multi(*region_entries, force: bust_cache?) { |region_entry|
+        earliest_patient_data_query_v2(region_entry.region)
+      }
+      cached_results.each_with_object({}) { |(region_entry, time), results| results[region_entry.slug] = time }
+    end
+
+    memoize def earliest_patient_recorded_at_period
+      earliest_patient_recorded_at.each_with_object({}) { |(slug, time), hsh| hsh[slug] = Period.new(value: time, type: @period_type) if time }
+    end
+
+    private def earliest_patient_data_query_v2(region)
+      FacilityState.for_region(region).where("cumulative_registrations > 0 OR cumulative_assigned_patients > 0")
+        .minimum(:month_date)
+    end
+
+    # Adjusted patient counts are the patient counts from three months ago (the adjusted period) that
+    # are the basis for control rates. These counts DO include lost to follow up.
+    memoize def adjusted_patients_with_ltfu
+      cumulative_assigned_patients.each_with_object({}) do |(entry, result), results|
+        values = periods.each_with_object(Hash.new(0)) { |period, region_result|
+          region_result[period] = result[period.adjusted_period]
+        }
+        results[entry] = values
+      end
+    end
+
+    # Adjusted patient counts are the patient counts from three months ago (the adjusted period) that
+    # are the basis for control rates. These counts DO NOT include lost to follow up.
+    memoize def adjusted_patients_without_ltfu
+      regions.each_with_object({}) { |region, result|
+        result[region.slug] = adjusted_patients_without_ltfu_query(region).slice(*periods.entries)
```

Comment:
Should we be making use of `period.adjusted_period` here? I don't see where the 3 month shift is happening.